packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -756,6 +945,148 @@ class RequireNotEqualGadget : public GadgetT
     }
 };
 
+// (A <(=) B)
+class LeqGadget : public GadgetT
+{
+public:
+    VariableT _lt;
+    VariableT _leq;
+    libsnark::comparison_gadget<ethsnarks::FieldT> comparison;
+    NotGadget _gt;
+    NotGadget _gte;
+    AndGadget _eq;
+
+    LeqGadget(
+        ProtoboardT& pb,
+        const VariableT& A,
+        const VariableT& B,
+        const size_t n,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+
+        _lt(make_variable(pb, 1, FMT(prefix, ".lt"))),
+        _leq(make_variable(pb, 1, FMT(prefix, ".leq"))),
+        comparison(pb, n, A, B, _lt, _leq, FMT(prefix, ".A <(=) B")),
+        _gt(pb, _leq, FMT(prefix, ".gt")),
+        _gte(pb, _lt, FMT(prefix, ".gte")),
+        _eq(pb, {_leq, _gte.result()}, FMT(prefix, ".eq"))
+    {
+        // The comparison gadget is only guaranteed to work correctly on values in the field capacity - 1
+        assert(n <= NUM_BITS_FIELD_CAPACITY - 1);
+    }
+
+    const VariableT& lt() const
+    {
+        return _lt;
+    }
+
+    const VariableT& leq() const
+    {
+        return _leq;
+    }
+
+    const VariableT& eq() const
+    {
+        return _eq.result();
+    }
+
+    const VariableT& gte() const
+    {
+        return _gte.result();
+    }
+
+    const VariableT& gt() const
+    {
+        return _gt.result();
+    }
+
+    void generate_r1cs_witness()
+    {
+        comparison.generate_r1cs_witness();
+        _gt.generate_r1cs_witness();
+        _gte.generate_r1cs_witness();
+        _eq.generate_r1cs_witness();
+    }
+
+    void generate_r1cs_constraints()
+    {
+        comparison.generate_r1cs_constraints();
+        _gt.generate_r1cs_constraints();
+        _gte.generate_r1cs_constraints();
+        _eq.generate_r1cs_constraints();
+    }
+};
+
+// (A < B)
```

Comment:
Comparison gadgets don't work correctly on values that can take up the full field value, because they use a trick where an extra bit is needed for it to work correctly. So if the value can already use up all the bits it won't work correctly and is totally not safe to use.

This gadget can work on values that take up the full field range by breaking up the comparison in 2 parts so the values are smaller and we can correctly use the comparison gadget.